timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java

```diff
@@ -64,6 +76,17 @@ public synchronized <T> ValueAndLockWatchStateUpdate<T> runTask(
     @Override
     public synchronized void logLock(
             Set<LockDescriptor> locksTakenOut, LockToken lockToken, Optional<LockRequestMetadata> metadata) {
+        // We don't want metadata to blow to consume absurd amounts of memory, but we cannot fail here as not publishing
+        // the LockEvent would mean cache corruption. Discarding the metadata is fine, but we should keep track of
+        // how often this happens.
+        int metadataSize = metadata.map(LockRequestMetadata::size).orElse(0);
+        if (metadataSize > MAX_METADATA_SIZE) {
+            log.warn(
+                    "Discarding metadata because it is too large",
+                    SafeArg.of("metadataSize", metadataSize),
+                    SafeArg.of("maxMetadataSize", MAX_METADATA_SIZE));
+            metadata = Optional.empty();
```

Comment:
The alternative is to shrink the metadata to `MAX_METADATA_SIZE`. However, this adds additional processing overhead and the decision which values to keep and which to discard is non-trivial.